testing/sdk_repository.py

```diff
@@ -96,18 +100,194 @@ def remove_universe_repos(stub_urls):
+        headers = {
+            "User-Agent": "dcos/{}".format(sdk_cmd.dcos_version()),
+            "Accept": "application/vnd.dcos.universe.repo+json;"
+                      "charset=utf-8;version={}".format('v4'),
+        }
+        req = urllib.request.Request(repo_url, headers=headers)
+        with urllib.request.urlopen(req) as f:
+            data = json.loads(f.read().decode())
+            for package in data['packages']:
+                package_file_paths.append(
+                    build_dcos_file_from_universe_definition(
+                        package,
+                        dcos_files_path,
+                        temp_dir
+                    )
+                )
+    return package_file_paths
+
+
+def build_dcos_file_from_universe_definition(
+        package: dict,
+        dcos_files_path: str,
+        temp_dir
+) -> str:
+    # TODO Ideally we should `migrate` and then `build`.
+    del package['releaseVersion']
+    del package['selected']
+    package_json_file = temp_dir.mkdtemp().join(random_string())
+    package_json_file.write(json.dumps(package))
+    target = os.path.join(
+        dcos_files_path,
+        '{}-{}.dcos'.format(package['name'], package['version'])
+    )
+    if os.path.isfile(target):
+        log.info('Skipping build, using cached file : {}'.format(target))
+    else:
+        rc, _, _ = sdk_cmd.run_raw_cli(' '.join([
+            'registry',
+            'build',
+            '--build-definition-file={}'.format(str(package_json_file)),
+            '--output-directory={}'.format(dcos_files_path),
+            '--json'
+        ]))
+        assert rc == 0
+    assert os.path.isfile(target)
+    return target
+
+
+def grant_perms_for_registry_account(service_uid: str) -> None:
+    # Grant only required permissions to registry
+    rc, _, _ = sdk_cmd.run_raw_cli(
+        "security org users grant {} '{}' '{}'".format(
+            service_uid,
+            'dcos:adminrouter:ops:ca:rw',
+            'full'
+        )
+    )
+    assert rc == 0
+
+
+@retry(
+    stop_max_delay=5 * 60 * 1000,
+    wait_fixed=5 * 1000
+)
+def loop_until_cli_condition(
+        raw_cli_cmd: str,
+        check: Callable[[int, str, str], bool],
+) -> None:
+    code, stdout, stderr = sdk_cmd.run_raw_cli(raw_cli_cmd)
+    assert check(code, stdout, stderr)
+
+
+def universe_session(temp_dir):
     """Add the universe package repositories defined in $STUB_UNIVERSE_URL.
 
     This should generally be used as a fixture in a framework's conftest.py:
 
     @pytest.fixture(scope='session')
-    def configure_universe():
-        yield from sdk_repository.universe_session()
+    def configure_universe(temp_dir):
+        yield from sdk_repository.universe_session(temp_dir)
     """
     stub_urls = {}
+    pkg_reg = {}
+    package_registry_enabled = os.environ['PACKAGE_REGISTRY_ENABLED'] == 'true'
     try:
-        stub_urls = add_universe_repos()
+        if package_registry_enabled:
```

Comment:
We could move the package registry specifics to their own function and also use the `try-yield-finally` semantics there. That would keep this function much simpler. for example:
```
if package_registry_enabled:
    yield sdk_package_registry.setup_package_registry()
else:
     try:
         stub_urls = add_stub_universe_urls(get_universe_repos())
         yield
      finally:
         remove_universe_repos(stub_urls)
```